File: gateway/internal/proto/proto.go
// Package proto implements just enough of the Postgres wire protocol to route:
// initial-packet framing, SSL/GSS negotiation codes, StartupMessage params, and
// ErrorResponse building. The gateway never speaks the query protocol; after
// startup it pipes bytes.
package proto

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
)

// Wire-protocol magic numbers.
const (
	SSLRequestCode    uint32 = 80877103
	GSSEncRequestCode uint32 = 80877104
	CancelRequestCode uint32 = 80877102
	Protocol30        uint32 = 196608 // 3 << 16
)

// Message types produced by ParseInitialPacket.
const (
	TypeSSL     = "ssl"
	TypeGSSEnc  = "gssenc"
	TypeCancel  = "cancel"
	TypeStartup = "startup"
)

// Msg is a classified initial packet.
type Msg struct {
	Type   string
	Params map[string]string
	Raw    []byte
}

// ReadInitialPacket frames a buffer as: int32 length (includes itself) + payload.
// It returns ok=false (with no error) when buf does not yet hold a complete
// packet, and an error for a bogus length. On success it returns the complete
// packet and any trailing bytes (rest) that followed it.
func ReadInitialPacket(buf []byte) (packet, rest []byte, ok bool, err error) {
	if len(buf) < 4 {
		return nil, nil, false, nil
	}
	length := int(binary.BigEndian.Uint32(buf[0:4]))
	if length < 8 || length > 10000 {
		return nil, nil, false, fmt.Errorf("bogus initial packet length %d", length)
	}
	if len(buf) < length {
		return nil, nil, false, nil
	}
	return buf[:length], buf[length:], true, nil
}

// ParseInitialPacket classifies and parses a complete initial packet.
func ParseInitialPacket(packet []byte) (Msg, error) {
	if len(packet) < 8 {
		return Msg{}, fmt.Errorf("initial packet too short: %d bytes", len(packet))
	}
	code := binary.BigEndian.Uint32(packet[4:8])
	switch code {
	case SSLRequestCode:
		return Msg{Type: TypeSSL, Raw: packet}, nil
	case GSSEncRequestCode:
		return Msg{Type: TypeGSSEnc, Raw: packet}, nil
	case CancelRequestCode:
		return Msg{Type: TypeCancel, Raw: packet}, nil
	}
	if code != Protocol30 {
		return Msg{}, fmt.Errorf("unsupported protocol %d.%d", code>>16, code&0xffff)
	}
	params := map[string]string{}
	off := 8
	for off < len(packet)-1 {
		kRel := bytes.IndexByte(packet[off:], 0)
		if kRel == -1 || kRel == 0 { // no key, or empty key = terminator
			break
		}
		kEnd := off + kRel
		vRel := bytes.IndexByte(packet[kEnd+1:], 0)
		if vRel == -1 {
			return Msg{}, fmt.Errorf("unterminated startup parameter")
		}
		vEnd := kEnd + 1 + vRel
		params[string(packet[off:kEnd])] = string(packet[kEnd+1 : vEnd])
		off = vEnd + 1
	}
	return Msg{Type: TypeStartup, Params: params, Raw: packet}, nil
}

// BuildStartup builds a StartupMessage from params (for tests and health probes).
func BuildStartup(params map[string]string) []byte {
	var body bytes.Buffer
	for k, v := range params {
		body.WriteString(k)
		body.WriteByte(0)
		body.WriteString(v)
		body.WriteByte(0)
	}
	body.WriteByte(0)
	out := make([]byte, 8+body.Len())
	binary.BigEndian.PutUint32(out[0:4], uint32(8+body.Len()))
	binary.BigEndian.PutUint32(out[4:8], Protocol30)
	copy(out[8:], body.Bytes())
	return out
}

// BuildSSLRequest builds an 8-byte SSLRequest packet.
func BuildSSLRequest() []byte {
	b := make([]byte, 8)
	binary.BigEndian.PutUint32(b[0:4], 8)
	binary.BigEndian.PutUint32(b[4:8], SSLRequestCode)
	return b
}

// BuildErrorResponse builds an ErrorResponse the client understands, so failures
// are visible in psql instead of a bare connection reset.
func BuildErrorResponse(code, message string) []byte {
	var body bytes.Buffer
	fields := [][2]string{{"S", "FATAL"}, {"V", "FATAL"}, {"C", code}, {"M", message}}
	for _, f := range fields {
		body.WriteString(f[0])
		body.WriteString(f[1])
		body.WriteByte(0)
	}
	body.WriteByte(0)
	out := make([]byte, 5+body.Len())
	out[0] = 'E'
	binary.BigEndian.PutUint32(out[1:5], uint32(4+body.Len()))
	copy(out[5:], body.Bytes())
	return out
}

// ReadBackendMessage reads one regular-protocol backend message (type byte +
// int32 length + body) from r. It returns the message type and the raw bytes
// exactly as read, so callers can forward them verbatim.
func ReadBackendMessage(r io.Reader) (typ byte, raw []byte, err error) {
	head := make([]byte, 5)
	if _, err = io.ReadFull(r, head); err != nil {
		return 0, nil, err
	}
	n := int(int32(binary.BigEndian.Uint32(head[1:5])))
	if n < 4 || n > 1<<20 {
		return head[0], head, fmt.Errorf("bogus backend message length %d", n)
	}
	body := make([]byte, n-4)
	if _, err = io.ReadFull(r, body); err != nil {
		return head[0], head, err
	}
	return head[0], append(head, body...), nil
}

// ErrorCode extracts the SQLSTATE ('C' field) from a raw ErrorResponse
// message as returned by ReadBackendMessage. Empty if absent.
func ErrorCode(raw []byte) string {
	if len(raw) < 6 || raw[0] != 'E' {
		return ""
	}
	body := raw[5:]
	for len(body) > 0 && body[0] != 0 {
		f := body[0]
		end := bytes.IndexByte(body[1:], 0)
		if end < 0 {
			return ""
		}
		if f == 'C' {
			return string(body[1 : 1+end])
		}
		body = body[2+end:]
	}
	return ""
}

File: gateway/src/proto.js
'use strict';
// Postgres wire protocol — just enough to route: initial-packet framing,
// SSL/GSS negotiation codes, StartupMessage params, ErrorResponse building.
// The gateway never speaks the query protocol; after startup it pipes bytes.

const SSL_REQUEST_CODE = 80877103;
const GSSENC_REQUEST_CODE = 80877104;
const CANCEL_REQUEST_CODE = 80877102;
const PROTOCOL_3_0 = 196608; // 3 << 16

// Initial packets are framed as: int32 length (includes itself) + payload.
// Returns null if buf doesn't yet hold a complete packet.
function readInitialPacket(buf) {
  if (buf.length < 4) return null;
  const len = buf.readInt32BE(0);
  if (len < 8 || len > 10000) throw new Error(`bogus initial packet length ${len}`);
  if (buf.length < len) return null;
  return { len, packet: buf.subarray(0, len), rest: buf.subarray(len) };
}

// Classify + parse a complete initial packet.
// -> { type: 'ssl'|'gssenc'|'cancel'|'startup', params?, raw }
function parseInitialPacket(packet) {
  const code = packet.readInt32BE(4);
  if (code === SSL_REQUEST_CODE) return { type: 'ssl', raw: packet };
  if (code === GSSENC_REQUEST_CODE) return { type: 'gssenc', raw: packet };
  if (code === CANCEL_REQUEST_CODE) return { type: 'cancel', raw: packet };
  if (code !== PROTOCOL_3_0) throw new Error(`unsupported protocol ${code >> 16}.${code & 0xffff}`);
  const params = {};
  let off = 8;
  while (off < packet.length - 1) {
    const kEnd = packet.indexOf(0, off);
    if (kEnd === -1 || kEnd === off) break; // empty key = terminator
    const vEnd = packet.indexOf(0, kEnd + 1);
    if (vEnd === -1) throw new Error('unterminated startup parameter');
    params[packet.toString('utf8', off, kEnd)] = packet.toString('utf8', kEnd + 1, vEnd);
    off = vEnd + 1;
  }
  return { type: 'startup', params, raw: packet };
}

// Build a StartupMessage from params (for tests and health probes).
function buildStartup(params) {
  const parts = [];
  for (const [k, v] of Object.entries(params)) {
    parts.push(Buffer.from(k, 'utf8'), Buffer.from([0]), Buffer.from(v, 'utf8'), Buffer.from([0]));
  }
  parts.push(Buffer.from([0]));
  const body = Buffer.concat(parts);
  const head = Buffer.alloc(8);
  head.writeInt32BE(8 + body.length, 0);
  head.writeInt32BE(PROTOCOL_3_0, 4);
  return Buffer.concat([head, body]);
}

function buildSslRequest() {
  const b = Buffer.alloc(8);
  b.writeInt32BE(8, 0);
  b.writeInt32BE(SSL_REQUEST_CODE, 4);
  return b;
}

// ErrorResponse the client understands, so failures are visible in psql
// instead of a bare connection reset.
function buildErrorResponse(code, message) {
  const fields = [
    ['S', 'FATAL'], ['V', 'FATAL'], ['C', code], ['M', message],
  ];
  const parts = [];
  for (const [t, v] of fields) {
    parts.push(Buffer.from(t, 'utf8'), Buffer.from(v, 'utf8'), Buffer.from([0]));
  }
  parts.push(Buffer.from([0]));
  const body = Buffer.concat(parts);
  const msg = Buffer.alloc(5 + body.length);
  msg.write('E', 0);
  msg.writeInt32BE(4 + body.length, 1);
  body.copy(msg, 5);
  return msg;
}

module.exports = {
  SSL_REQUEST_CODE, GSSENC_REQUEST_CODE, CANCEL_REQUEST_CODE, PROTOCOL_3_0,
  readInitialPacket, parseInitialPacket, buildStartup, buildSslRequest, buildErrorResponse,
};
